[CBRD-26250] Replication option handling

youngjun9072 opened the pull request:
<http://jira.cubrid.org/browse/CBRD-26250>

### Purpose

테이블 생성(CREATE TABLE...) , 테이블 수정(ALTER TABLE ...) 에 추가된 REPLICATION 옵션 저장 및 조회 기능을 추가합니다.

### Implementation
1. REPLICATION 옵션 출력
추가된 CREATE TABLE..., ALTER TABLE... 구문은 sm_set_class_flag() 함수를 통해 클래스의 플래그를 설정하며, 플래그는 SM_CLASSFLAG_NO_REPLICATION 으로 정의됩니다.

단, 제약 조건에따라 HA모드에서 리플리케이션 옵션 off->on 은 사용될 수 없으며, "Replication setting is not permitted for this table." 에러를 발생시킵니다
에러 코드: ER_REPLICATION_CONSTRAINT( -1368) .

2. REPLICATION 옵션 출력
SELECT is_replication FROM db_class 와 SHOW CREATE TABLE 구문을 사용할 수 있습니다
이 구문으로는 복제 옵션을 확인할 수 있으며 이를 위해 describe_class 에서 해당 클래스의 SM_CLASSFLAG_NO_REPLICATION 플래그 설정 값에따라 복제 여부를 출력합니다.
 
테스트 가능한 구문은 jira 이슈에 포함되어 있으며, 특히 ha모드에서 off->on 시 에러 발생을 확인하기위해선 HA 환경을 구축하신 후 ALTER TABLE 을 테스트하시면됩니다.

### Remarks
옵션 명령어인 "REPLICATION" 에 따라 플래그명 또한 SM_CLASSFLAG_REPLICATION로 명명하고, 사용자가 복제 테이블을 지정할 시 이 플래그를 TRUE로 설정하는 방법이 고려되었다. 하지만 하위버전에는 REPLICATION 정보가 존재하지 않아 하위버전의 백업 패치 이후버전으로 복원할 경우 모두 복제 제외테이블이 된다.
이러한 이유로 플래그 명은 SM_CLASSFLAG_NO_REPLICATION로 하고 복제 제외테이블에 대해서만 플래그를 설정하며, 이 플래그가 설정되어있지 않은 테이블(하위 버전 포함)은 모두 복제 대상이된다.

youngjun9072 committed:
add replication option

youngjun9072 committed:
lint

youngjun9072 committed:
fix typo

youngjun9072 committed:
Add variable intialization

youngjun9072 committed:
add "alter table ... replication" option

youngjun9072 committed:
lint

youngjun9072 committed:
add "TODO" comment

youngjun9072 committed:
add replication opt handling and select catalog

youngjun9072 committed:
Code changes

- Changed the location where replication option is handled
- Simplified condition logic to check only default or "on"
- Added explanatory comments

youngjun9072 committed:
lint

youngjun9072 committed:
replication 저장 및 조회 구현

youngjun9072 committed:
add alter table handling

youngjun9072 committed:
리뷰 반영
 - 매크로 이름 변경
 - 불필요 공백 삭제

youngjun9072 committed:
에러 핸들링 추가

youngjun9072 committed:
add tab

youngjun9072 committed:
add comment

youngjun9072 committed:
Flag and condition changes for backward compatibility with older backups

- replace SM_CLASSFLAG_REPLICATION with SM_CLASSFLAG_NO_REPLICATION.
- change handling condition from IS_REPLICATION_ON to !IS_REPLICATION.

youngjun9072 committed:
Add macro for alter condition branch

youngjun9072 committed:
Code modification

- Rename flag from SM_CLASSFLAG_REPLICATION to SM_CLASSFLAG_NO_REPLICATION
- Update return values to reflect the new flag naming
- Add error handling when attempting to alter replication tables in HA mode

youngjun9072 committed:
Add error message

youngjun9072 committed:
Merge branch 'feature/add-replication-option' into replication-option-handling

youngjun9072 committed:
remove unused error log

youngjun9072 committed:
Macro refactoring and lint

youngjun9072 committed:
Refactoring

youngjun9072 committed:
lint

H2SU commented on src/object/schema_system_catalog_install_query_spec.cpp:
```diff
@@ -70,6 +70,7 @@ sm_define_view_class_spec (void)
 	  "CAST ([c].[owner].[name] AS VARCHAR(255)) AS [owner_name], " /* string -> varchar(255) */
 	  "CASE [c].[class_type] WHEN 0 THEN 'CLASS' WHEN 1 THEN 'VCLASS' ELSE 'UNKNOWN' END AS [class_type], "
 	  "CASE WHEN MOD ([c].[is_system_class], 2) = 1 THEN 'YES' ELSE 'NO' END AS [is_system_class], "
+          "CASE WHEN MOD ([c].[is_system_class] / 32, 2) = 1 THEN 'NO' ELSE 'YES' END AS [is_replication_class], "
```
indent가 안 맞는 것 같습니다!

youngjun9072 replied:
내용 반영했습니다.

hgryoo commented on src/object/class_object.h:
```diff
@@ -308,7 +308,8 @@ typedef enum
   SM_CLASSFLAG_REUSE_OID = 8,	/* the class can reuse OIDs */
-  SM_CLASSFLAG_SUPPLEMENTAL_LOG = 16	/* reserved flag for supplemental log. */
+  SM_CLASSFLAG_SUPPLEMENTAL_LOG = 16,	/* reserved flag for supplemental log. */
+  SM_CLASSFLAG_NO_REPLICATION = 32	/* class is replicable */
```
```suggestion
  SM_CLASSFLAG_NO_REPLICATION = 32	/* class is not replicable */
```
nit

hgryoo replied:
이전 코드에서 IS_REPLICATION_ON_NODE 와 같은 매크로가 있는 것을 보았을 때
SM_CLASSFLAG_REPLICATION_OFF 와 같이 표현을 통일하는것도 좋아보입니다.

youngjun9072 replied:
플러그 이름 변경됐습니다. 추가로 주석의 내용도 변경됐고, PR 공지에 언급한것처럼 스키마는 복제하지만 데이터는 복제하지 않는다는 내용을 강조했습니다.

hgryoo commented on src/object/schema_manager.c:
```diff
@@ -3385,6 +3385,26 @@ sm_check_reuse_oid_class (MOP op)
+ * op(in): class object
+ */
+bool
+sm_is_replication_class (MOP op)
```
```suggestion
sm_is_replicable_class (MOP op)
```
복제 가능한 클래스인지 확인하는 것 같아, 이름 변경 제안 드립니다.

youngjun9072 replied:
이거에대해 고려해봤는데 스펙문서상에서도 "복제 테이블" 이라는 표현을 썼어서 일단은 유지하려합니다.

hornetmj commented on src/base/error_code.h:
```diff
@@ -1750,7 +1750,9 @@
-#define ER_LAST_ERROR                               -1368
+#define ER_REPLICATION_CONSTRAINT                   -1368
+
+#define ER_LAST_ERROR                               -1369
```
에러 번호 정렬

youngjun9072 replied:
반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -10664,28 +10667,104 @@ do_alter_change_col_comment (PARSER_CONTEXT * const parser, PT_NODE * const alte
+  class_mop = ctemplate->op;
+  replication_node = alter->info.alter.alter_clause.replication.tbl_replication;
+
+  if (!HA_DISABLED () && IS_REPLICATION_ON_NODE (replication_node))
```
전달 받은 정보로 체크할 수 있는 부분으로 함수 엔트리에서 체크되면 불필요한 작업들이 수행될 필요가 없어집니다.

youngjun9072 replied:
반영했습니다. 추가로 주석의 위치가 실행코드의 바로 위에서 함수의 위로 변경됐습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -10664,28 +10667,104 @@ do_alter_change_col_comment (PARSER_CONTEXT * const parser, PT_NODE * const alte
-      _er_log_debug (ARG_FILE_LINE, "[Not implemented] %s(replication=on) table replication set to on. \n",
-		     entity_name);
+      error = ER_REPLICATION_CONSTRAINT;
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 1, "Replication setting is not permitted for this table.");
```
1. ER_REPLICATION_CONSTRAINT 에러 메시지에는 인자가 없는데, 인자 개수를 의미하는 1과 인자값 "Replication setting is not permitted for this table."가 전달되고 있습니다.
2. 에러 메시지 변경 검토 요청 'Changing the replication option of a class is not allowed in HA mode.'

youngjun9072 replied:
반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -10664,28 +10667,104 @@ do_alter_change_col_comment (PARSER_CONTEXT * const parser, PT_NODE * const alte
     }
-  else
+
+  error = sm_set_class_flag (class_mop, SM_CLASSFLAG_NO_REPLICATION, !IS_REPLICATION_ON_NODE (replication_node));
```
error 핸들링 안됨

youngjun9072 replied:
반영했습니다.

youngjun9072 committed:
Apply review feedback
- fix indentation
- rename SM_CLASSFLAG_NO_REPLICATION to SM_CLASSFLAG_REPLICATION_OFF and update comment
- check replication status at function entry to avoid unnecessary operations
- adjust error message and parameter from 1 to 0
- add error handling

hornetmj commented on src/object/class_object.h:
```diff
@@ -308,7 +308,8 @@ typedef enum
   SM_CLASSFLAG_REUSE_OID = 8,	/* the class can reuse OIDs */
-  SM_CLASSFLAG_SUPPLEMENTAL_LOG = 16	/* reserved flag for supplemental log. */
+  SM_CLASSFLAG_SUPPLEMENTAL_LOG = 16,	/* reserved flag for supplemental log. */
+  SM_CLASSFLAG_REPLICATION_OFF = 32	/* disables replication of class data (schema still replicated) */
```
```suggestion
  SM_CLASSFLAG_REPLICATION_DATA_OFF = 32	/* disables replication of class data (schema still replicated) */
```

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -10663,29 +10666,114 @@ do_alter_change_col_comment (PARSER_CONTEXT * const parser, PT_NODE * const alte
-   * the default value is set to "on".
-   */
-  if (IS_REPLICATION_ON_NODE (replication_node))
+  if (!HA_DISABLED () && IS_REPLICATION_ON_NODE (replication_node))
```
- HA_MODE에서 ON -> OFF는 허용되는 것인가요? 추후 FK와 엮이면 복잡해질 것 같아 허용하지 않도록 정리되었던 것 같아 문의드립니다.
- 매크로 명칭에 NODE라는 표현이 PT_NODE에서 파생된 것 같은데 HA의 master node, slave node와 같이 고려되어 오해의 소지가 발생할 수 있을 것 같습니다. OPTION 등의 표현을 사용하시면 어떨까 합니다.
